File: typescript/12-multi-agente/equipo_roles.ts
// Patrón Equipo de Roles (MetaGPT style): PM → Architect → ProjectManager → Engineer → QA
// Coordinados por un Message Pool con campo cause_by. Engineer tiene 3 reintentos.

// Cómo ejecutar: make ts SCRIPT=typescript/12-multi-agente/equipo_roles.ts

import Anthropic from "@anthropic-ai/sdk";

const client = new Anthropic();
const MODEL = process.env["MODEL"] ?? "claude-haiku-4-5-20251001";
const MAX_ENGINEER_RETRIES = 3;

// ---- Message Pool ----

interface Message {
  content: string;
  type: string;
  causeBy: string;
}

class MessagePool {
  private messages: Message[] = [];

  publish(content: string, type: string, causeBy: string): void {
    this.messages.push({ content, type, causeBy });
  }

  latest(type: string): Message | undefined {
    return [...this.messages].reverse().find(m => m.type === type);
  }

  summary(): string {
    return this.messages
      .map(m => `  [${m.causeBy}] ${m.type}: ${m.content.slice(0, 60)}...`)
      .join("\n");
  }
}

// ---- Roles ----

const ROLES = {
  pm: {
    id: "pm",
    systemPrompt:
      "Eres un Product Manager senior. A partir del requisito del usuario, " +
      "escribe un PRD claro y estructurado. Incluye: objetivo, funcionalidades, " +
      "criterios de aceptación y restricciones técnicas. Sé específico.",
  },
  architect: {
    id: "architect",
    systemPrompt:
      "Eres un Arquitecto de Software. A partir del PRD, diseña la arquitectura. " +
      "Incluye: componentes principales, interfaces, decisiones de diseño (con justificación) " +
      "y lista de archivos/módulos a crear.",
  },
  pm_mgr: {
    id: "pm_mgr",
    systemPrompt:
      "Eres un Project Manager técnico. A partir del System Design, " +
      "crea un plan de implementación serializable con tareas y dependencias explícitas. " +
      "El ingeniero ejecutará este plan directamente.",
  },
  engineer: {
    id: "engineer",
    systemPrompt:
      "Eres un Ingeniero de Software senior. A partir del plan, " +
      "escribe el código funcional completo. Incluye todos los archivos necesarios " +
      "con sintaxis correcta. El QA ejecutará tests — asegúrate de que sea ejecutable.",
  },
  qa: {
    id: "qa",
    systemPrompt:
      "Eres un QA Engineer. Revisa el código contra el PRD y el System Design. " +
      "Busca: bugs de lógica, casos borde no cubiertos, violaciones de requisitos. " +
      'Responde con JSON: {"tiene_bugs": true/false, "bugs": [...], "veredicto": "PASA" o "FALLA"}.',
  },
};

// ---- Helper ----

async function llamarLLM(system: string, user: string, temperature = 0.0): Promise<string> {
  const resp = await client.messages.create({
    model: MODEL,
    max_tokens: 1200,
    system,
    messages: [{ role: "user", content: user }],
    temperature,
  });
  return (resp.content[0] as { type: "text"; text: string }).text.trim();
}

function parsearTestReport(raw: string): { tiene_bugs: boolean; bugs: string[]; veredicto: string } {
  const inicio = raw.indexOf("{");
  const fin = raw.lastIndexOf("}") + 1;
  if (inicio === -1 || fin === 0) return { tiene_bugs: false, bugs: [], veredicto: "PASA" };
  try {
    return JSON.parse(raw.slice(inicio, fin));
  } catch {
    return { tiene_bugs: false, bugs: [], veredicto: "PASA" };
  }
}

// ---- Pipeline principal ----

async function equipoRoles(requisitoUsuario: string): Promise<string> {
  const pool = new MessagePool();

  // PM: genera PRD
  console.log("[PM] Generando PRD...");
  const prd = await llamarLLM(ROLES.pm.systemPrompt, `Requisito del usuario: ${requisitoUsuario}`);
  pool.publish(prd, "PRD", "pm");
  console.log(`  PRD: ${prd.slice(0, 80)}...`);

  // Architect: observa PM → SystemDesign
  console.log("\n[Architect] Diseñando sistema...");
  const prdMsg = pool.latest("PRD")!;
  const systemDesign = await llamarLLM(ROLES.architect.systemPrompt, `PRD:\n${prdMsg.content}`);
  pool.publish(systemDesign, "SystemDesign", "architect");
  console.log(`  SystemDesign: ${systemDesign.slice(0, 80)}...`);

  // PM_Manager: observa Architect → Plan
  console.log("\n[ProjectManager] Creando plan...");
  const archMsg = pool.latest("SystemDesign")!;
  const plan = await llamarLLM(
    ROLES.pm_mgr.systemPrompt,
    `PRD:\n${prdMsg.content}\n\nSystem Design:\n${archMsg.content}`
  );
  pool.publish(plan, "Plan", "pm_mgr");
  console.log(`  Plan: ${plan.slice(0, 80)}...`);

  // Engineer: observa PM_Manager → Code
  console.log("\n[Engineer] Escribiendo código...");
  const planMsg = pool.latest("Plan")!;
  const contextoEng = `PRD:\n${prdMsg.content}\n\nSystem Design:\n${archMsg.content}\n\nPlan:\n${planMsg.content}`;
  let codigo = await llamarLLM(ROLES.engineer.systemPrompt, contextoEng, 0.2);
  pool.publish(codigo, "Code", "engineer");
  console.log(`  Code: ${codigo.slice(0, 80)}...`);

  // QA: revisión inicial
  console.log("\n[QA] Revisando código...");
  const codeMsg = pool.latest("Code")!;
  const qaContexto = `PRD:\n${prdMsg.content}\n\nSystem Design:\n${archMsg.content}\n\nCódigo:\n${codeMsg.content}`;
  let testReportRaw = await llamarLLM(ROLES.qa.systemPrompt, qaContexto);
  pool.publish(testReportRaw, "TestReport", "qa");
  let reporte = parsearTestReport(testReportRaw);
  console.log(`  Veredicto inicial: ${reporte.veredicto}`);

  // Bucle Engineer ↔ QA: hasta MAX_ENGINEER_RETRIES
  let intento = 0;
  while (reporte.tiene_bugs && intento < MAX_ENGINEER_RETRIES) {
    intento++;
    const bugsTexto = reporte.bugs.map(b => `- ${b}`).join("\n");
    console.log(`\n[Engineer] Intento ${intento}/${MAX_ENGINEER_RETRIES} — corrigiendo bugs:\n  ${bugsTexto.slice(0, 120)}`);

    const codigoActual = pool.latest("Code") ?? pool.latest("BugFix");
    const fix = await llamarLLM(
      ROLES.engineer.systemPrompt,
      `QA encontró bugs:\n${bugsTexto}\n\nCódigo actual:\n${codigoActual?.content}\n\nPRD:\n${prdMsg.content}\n\nCorrige todos los bugs.`,
      0.2
    );
    pool.publish(fix, "BugFix", "engineer");

    console.log(`\n[QA] Re-revisando (intento ${intento})...`);
    testReportRaw = await llamarLLM(
      ROLES.qa.systemPrompt,
      `PRD:\n${prdMsg.content}\n\nSystem Design:\n${archMsg.content}\n\nCódigo corregido:\n${fix}`
    );
    pool.publish(testReportRaw, "TestReport", "qa");
    reporte = parsearTestReport(testReportRaw);
    console.log(`  Veredicto intento ${intento}: ${reporte.veredicto}`);
  }

  const codigoFinal = pool.latest("BugFix") ?? pool.latest("Code");
  const estado = !reporte.tiene_bugs ? "VALIDADO" : "MEJOR_INTENTO";
  console.log(`\n[Pipeline] Estado final: ${estado}`);
  console.log(`\n--- Pool de mensajes ---\n${pool.summary()}`);

  return codigoFinal?.content ?? "[Sin código generado]";
}

async function main() {
  const requisito =
    "Implementa una función Python que calcule el número de Fibonacci de forma eficiente " +
    "usando memoización. Debe manejar n=0, n=1, y números grandes. " +
    "Incluye una función main() con ejemplos de uso.";
  console.log(`Requisito: ${requisito}\n${"=".repeat(60)}\n`);
  const resultado = await equipoRoles(requisito);
  console.log(`\n${"=".repeat(60)}\nCódigo final:\n${resultado}`);
}

main().catch(console.error);

File: go/12-multi-agente/equipo_roles.go
// Patrón Equipo de Roles (MetaGPT style): PM → Architect → ProjectManager → Engineer → QA
// Coordinados por un Message Pool con campo cause_by. Engineer tiene 3 reintentos.

// Cómo ejecutar: make go FILE=go/12-multi-agente/equipo_roles.go

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
)

var modelEquipo = envOr("SMALL_MODEL", "claude-haiku-4-5-20251001")
const maxEngineerRetries = 3

// ---- Message Pool ----

type PoolMessage struct {
	Content string
	Type    string
	CauseBy string
}

type MessagePool struct {
	messages []PoolMessage
}

func (p *MessagePool) Publish(content, tipo, causeBy string) {
	p.messages = append(p.messages, PoolMessage{Content: content, Type: tipo, CauseBy: causeBy})
}

func (p *MessagePool) Latest(tipo string) *PoolMessage {
	for i := len(p.messages) - 1; i >= 0; i-- {
		if p.messages[i].Type == tipo {
			return &p.messages[i]
		}
	}
	return nil
}

func (p *MessagePool) Summary() string {
	var sb strings.Builder
	for _, m := range p.messages {
		content := m.Content
		if len(content) > 60 {
			content = content[:60]
		}
		sb.WriteString(fmt.Sprintf("  [%s] %s: %s...\n", m.CauseBy, m.Type, content))
	}
	return sb.String()
}

// ---- Roles ----

type RoleSpec struct {
	ID           string
	SystemPrompt string
}

var roleSpecs = map[string]RoleSpec{
	"pm": {
		ID: "pm",
		SystemPrompt: "Eres un Product Manager senior. A partir del requisito del usuario, " +
			"escribe un PRD claro y estructurado. Incluye: objetivo, funcionalidades, " +
			"criterios de aceptación y restricciones técnicas. Sé específico.",
	},
	"architect": {
		ID: "architect",
		SystemPrompt: "Eres un Arquitecto de Software. A partir del PRD, diseña la arquitectura. " +
			"Incluye: componentes principales, interfaces, decisiones de diseño (con justificación) " +
			"y lista de archivos/módulos a crear.",
	},
	"pm_mgr": {
		ID: "pm_mgr",
		SystemPrompt: "Eres un Project Manager técnico. A partir del System Design, " +
			"crea un plan de implementación serializable con tareas y dependencias explícitas. " +
			"El ingeniero ejecutará este plan directamente.",
	},
	"engineer": {
		ID: "engineer",
		SystemPrompt: "Eres un Ingeniero de Software senior. A partir del plan, " +
			"escribe el código funcional completo. Incluye todos los archivos necesarios " +
			"con sintaxis correcta. El QA ejecutará tests — asegúrate de que sea ejecutable.",
	},
	"qa": {
		ID: "qa",
		SystemPrompt: "Eres un QA Engineer. Revisa el código contra el PRD y el System Design. " +
			"Busca: bugs de lógica, casos borde no cubiertos, violaciones de requisitos. " +
			`Responde con JSON: {"tiene_bugs": true/false, "bugs": [...], "veredicto": "PASA" o "FALLA"}.`,
	},
}

// ---- Helper ----

func llamarLLMEquipo(system, user string, temperature float64) (string, error) {
	type Msg struct {
		Role    string `json:"role"`
		Content string `json:"content"`
	}
	body, _ := json.Marshal(map[string]interface{}{
		"model":       modelEquipo,
		"max_tokens":  1200,
		"system":      system,
		"messages":    []Msg{{Role: "user", Content: user}},
		"temperature": temperature,
	})
	req, _ := http.NewRequest("POST", envBaseURL(), bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", os.Getenv("ANTHROPIC_API_KEY"))
	req.Header.Set("anthropic-version", "2023-06-01")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	var result struct {
		Content []struct {
			Type string `json:"type"`
			Text string `json:"text"`
		} `json:"content"`
	}
	json.Unmarshal(respBody, &result)
	if len(result.Content) > 0 {
		return strings.TrimSpace(result.Content[0].Text), nil
	}
	return "", fmt.Errorf("respuesta vacía")
}

type TestReport struct {
	TieneBugs bool     `json:"tiene_bugs"`
	Bugs      []string `json:"bugs"`
	Veredicto string   `json:"veredicto"`
}

func parsearTestReport(raw string) TestReport {
	inicio := strings.Index(raw, "{")
	fin := strings.LastIndex(raw, "}") + 1
	if inicio == -1 || fin == 0 {
		return TestReport{TieneBugs: false, Bugs: nil, Veredicto: "PASA"}
	}
	var report TestReport
	if err := json.Unmarshal([]byte(raw[inicio:fin]), &report); err != nil {
		return TestReport{TieneBugs: false, Bugs: nil, Veredicto: "PASA"}
	}
	return report
}

func trunc(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

// ---- Pipeline principal ----

func equipoRoles(requisitoUsuario string) (string, error) {
	pool := &MessagePool{}

	// PM: genera PRD
	fmt.Println("[PM] Generando PRD...")
	prd, err := llamarLLMEquipo(roleSpecs["pm"].SystemPrompt, "Requisito del usuario: "+requisitoUsuario, 0.0)
	if err != nil {
		return "", err
	}
	pool.Publish(prd, "PRD", "pm")
	fmt.Printf("  PRD: %s...\n", trunc(prd, 80))

	// Architect: observa PM → SystemDesign
	fmt.Println("\n[Architect] Diseñando sistema...")
	prdMsg := pool.Latest("PRD")
	systemDesign, err := llamarLLMEquipo(roleSpecs["architect"].SystemPrompt, "PRD:\n"+prdMsg.Content, 0.0)
	if err != nil {
		return "", err
	}
	pool.Publish(systemDesign, "SystemDesign", "architect")
	fmt.Printf("  SystemDesign: %s...\n", trunc(systemDesign, 80))

	// PM_Manager: observa Architect → Plan
	fmt.Println("\n[ProjectManager] Creando plan...")
	archMsg := pool.Latest("SystemDesign")
	plan, err := llamarLLMEquipo(
		roleSpecs["pm_mgr"].SystemPrompt,
		fmt.Sprintf("PRD:\n%s\n\nSystem Design:\n%s", prdMsg.Content, archMsg.Content),
		0.0,
	)
	if err != nil {
		return "", err
	}
	pool.Publish(plan, "Plan", "pm_mgr")
	fmt.Printf("  Plan: %s...\n", trunc(plan, 80))

	// Engineer: observa PM_Manager → Code
	fmt.Println("\n[Engineer] Escribiendo código...")
	planMsg := pool.Latest("Plan")
	contextoEng := fmt.Sprintf("PRD:\n%s\n\nSystem Design:\n%s\n\nPlan:\n%s",
		prdMsg.Content, archMsg.Content, planMsg.Content)
	codigo, err := llamarLLMEquipo(roleSpecs["engineer"].SystemPrompt, contextoEng, 0.2)
	if err != nil {
		return "", err
	}
	pool.Publish(codigo, "Code", "engineer")
	fmt.Printf("  Code: %s...\n", trunc(codigo, 80))

	// QA: revisión inicial
	fmt.Println("\n[QA] Revisando código...")
	codeMsg := pool.Latest("Code")
	qaContexto := fmt.Sprintf("PRD:\n%s\n\nSystem Design:\n%s\n\nCódigo:\n%s",
		prdMsg.Content, archMsg.Content, codeMsg.Content)
	testReportRaw, err := llamarLLMEquipo(roleSpecs["qa"].SystemPrompt, qaContexto, 0.0)
	if err != nil {
		return "", err
	}
	pool.Publish(testReportRaw, "TestReport", "qa")
	reporte := parsearTestReport(testReportRaw)
	fmt.Printf("  Veredicto inicial: %s\n", reporte.Veredicto)

	// Bucle Engineer ↔ QA: hasta maxEngineerRetries
	intento := 0
	for reporte.TieneBugs && intento < maxEngineerRetries {
		intento++
		var bugsParts []string
		for _, b := range reporte.Bugs {
			bugsParts = append(bugsParts, "- "+b)
		}
		bugsTexto := strings.Join(bugsParts, "\n")
		fmt.Printf("\n[Engineer] Intento %d/%d — corrigiendo bugs:\n  %s\n",
			intento, maxEngineerRetries, trunc(bugsTexto, 120))

		codigoActual := pool.Latest("BugFix")
		if codigoActual == nil {
			codigoActual = pool.Latest("Code")
		}
		fix, err := llamarLLMEquipo(
			roleSpecs["engineer"].SystemPrompt,
			fmt.Sprintf("QA encontró bugs:\n%s\n\nCódigo actual:\n%s\n\nPRD:\n%s\n\nCorrige todos los bugs.",
				bugsTexto, codigoActual.Content, prdMsg.Content),
			0.2,
		)
		if err != nil {
			return "", err
		}
		pool.Publish(fix, "BugFix", "engineer")

		fmt.Printf("\n[QA] Re-revisando (intento %d)...\n", intento)
		testReportRaw, err = llamarLLMEquipo(
			roleSpecs["qa"].SystemPrompt,
			fmt.Sprintf("PRD:\n%s\n\nSystem Design:\n%s\n\nCódigo corregido:\n%s",
				prdMsg.Content, archMsg.Content, fix),
			0.0,
		)
		if err != nil {
			return "", err
		}
		pool.Publish(testReportRaw, "TestReport", "qa")
		reporte = parsearTestReport(testReportRaw)
		fmt.Printf("  Veredicto intento %d: %s\n", intento, reporte.Veredicto)
	}

	codigoFinal := pool.Latest("BugFix")
	if codigoFinal == nil {
		codigoFinal = pool.Latest("Code")
	}
	estado := "VALIDADO"
	if reporte.TieneBugs {
		estado = "MEJOR_INTENTO"
	}
	fmt.Printf("\n[Pipeline] Estado final: %s\n", estado)
	fmt.Printf("\n--- Pool de mensajes ---\n%s\n", pool.Summary())

	if codigoFinal == nil {
		return "[Sin código generado]", nil
	}
	return codigoFinal.Content, nil
}

func main() {
	requisito := "Implementa una función Python que calcule el número de Fibonacci de forma eficiente " +
		"usando memoización. Debe manejar n=0, n=1, y números grandes. " +
		"Incluye una función main() con ejemplos de uso."
	fmt.Printf("Requisito: %s\n%s\n\n", requisito, strings.Repeat("=", 60))

	resultado, err := equipoRoles(requisito)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
		os.Exit(1)
	}
	fmt.Printf("\n%s\nCódigo final:\n%s\n", strings.Repeat("=", 60), resultado)
}

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func envBaseURL() string {
	if v := os.Getenv("ANTHROPIC_BASE_URL"); v != "" {
		return v + "/v1/messages"
	}
	return "https://api.anthropic.com/v1/messages"
}
